### PR TITLE
use HSTS when doing request with the HttpClient

### PR DESCRIPTION
**Pull request by rullzer**
Time to ensure that if a server we talk to supports HSTS we actually store this info and use it.

Help and commits to this branch are appreciated.

TODO
- [ ] cleanup ;)

Future work:
- [ ] Use preload list
- [ ] Extract HSTS logic a bit more (maybe add a public itnerface somewhere?)
- [ ] Use for other clients as well

### CODE REVIEW

File: lib/private/Http/Client/HSTSStore.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+		if ($maxAge <= 0) {
+			return;
+		}
+
+		$this->logger->warning("TIME TO SET HSTS");
+
+		$expires = $this->timeFactory->getTime() + $maxAge;
+
+		$qb = $this->db->getQueryBuilder();
```
Comment (CarlSchwan) [inline lib/private/Http/Client/HSTSStore.php:123]: could we maybe make hosts unique and if we get an unique exeption when trying to insert then we try updating. Save one DB query in some cases

Also we probably should try to catch data races

File: lib/private/Http/Client/HSTSMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	public function addHSTS() {
+		return function (callable $handler) {
+			return function (
+				RequestInterface $request,
+				array $options
+			) use ($handler) {
+
+                $request = $this->handleHSTSRewrite($request);
+
+                $this->logger->warning("GONNA REQUEST");
```
Comment (nickvergessen) [inline lib/private/Http/Client/HSTSMiddleware.php:92]: debug

File: lib/private/Http/Client/HSTSMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                $this->logger->warning($request->getUri()->getScheme());
+                $this->logger->warning($request->getUri()->getHost());
+
+
+				return $handler($request, $options)
+                    ->then(function (ResponseInterface $response) use ($request) {
+                        $this->logger->error("GOT RESPONSE");
```
Comment (nickvergessen) [inline lib/private/Http/Client/HSTSMiddleware.php:99]: debug

File: lib/private/Http/Client/HSTSStore.php (hunk 1 above)
Comment (rullzer) [reply lib/private/Http/Client/HSTSStore.php:123]: https://github.com/nextcloud/server/pull/34553/files#diff-07c117b72c7077964af8f2059557c552e79c4ef1b291d64cfece7b209204096aR74 it is unique

File: core/Migrations/Version26000Date20221011203714.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+ * Auto-generated migration step: Please modify to your needs!
+ */
+class Version26000Date20221011203714 extends SimpleMigrationStep {
+
+	/**
+	 * @param IOutput $output
+	 * @param Closure(): ISchemaWrapper $schemaClosure
+	 * @param array $options
+	 */
+	public function preSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
```
[the remark this answers is not in the document]
Comment (rullzer) [reply core/Migrations/Version26000Date20221011203714.php]: O joy :S

File: lib/private/Http/Client/HSTSMiddleware.php (hunk 2 above)
Comment (rullzer) [reply lib/private/Http/Client/HSTSMiddleware.php:92]: yeah plenty of those

**Pull request closed without merging**